### PR TITLE
[EDITORS-250] Improve filter for tree editors for UsageModel, SEFF and Assembly Diagrams

### PR DESCRIPTION
**Pull request by Yabbies**

### CODE REVIEW

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/AbstractActionItemProvider.java [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		super.addSuccessor_AbstractActionPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(AbstractAction.class, AbstractAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(AbstractAction object,
+                    List<AbstractAction> typedList) {
+            	ResourceDemandingBehaviour resourceDemandingBehaviour = object.getResourceDemandingBehaviour_AbstractAction();
+            	EList<AbstractAction> actionList = resourceDemandingBehaviour.getSteps_Behaviour();
+            	return actionList.stream()
+            			.filter(Predicate.not(StartAction.class::isInstance)) //EMF Package verwenden.
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/AbstractActionItemProvider.java:35]: As you commented already, use the SeffPackage.Literals.START_ACTION.isInstance method

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/AbstractActionItemProvider.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+		super.addPredecessor_AbstractActionPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(AbstractAction.class, AbstractAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(AbstractAction object,
+                    List<AbstractAction> typedList) {
+            	ResourceDemandingBehaviour resourceDemandingBehaviour = object.getResourceDemandingBehaviour_AbstractAction();
+            	EList<AbstractAction> actionList = resourceDemandingBehaviour.getSteps_Behaviour();
+            	return actionList.stream()
+            			.filter(Predicate.not(StopAction.class::isInstance))
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/AbstractActionItemProvider.java:53]: same here.

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/StartActionItemProvider.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	protected void addPredecessor_AbstractActionPropertyDescriptor(Object object) {
+		super.addPredecessor_AbstractActionPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(StartAction.class, AbstractAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(StartAction object,
+                    List<AbstractAction> typedList) {
+            	return new ArrayList<AbstractAction>();
+            }
+		});
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/StartActionItemProvider.java:34]: Move to either class or static utility method (e.g. createEmptyValueChoiceCalculator) to make things clear and reuse it for the StopAction.

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/StopActionItemProvider.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	protected void addSuccessor_AbstractActionPropertyDescriptor(Object object) {
+		super.addSuccessor_AbstractActionPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(StopAction.class, AbstractAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(StopAction object,
+                    List<AbstractAction> typedList) {
+            	return new ArrayList<AbstractAction>();
+            }
+		});
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/seff/provider/StopActionItemProvider.java:35]: Reuse from above

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+            						.stream()
+            						.filter(OperationRequiredRole.class::isInstance)
+            						.map(OperationRequiredRole.class::cast)
+            						.anyMatch(opr -> opr.getRequiredInterface__OperationRequiredRole().isAssignableFrom(oi))) {
+            					contexts2.add(c);
+            				}
+            			}
+            		}
+            		contexts.clear();
+            		contexts.addAll(contexts2);
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:87]: Please fix the indentation. It got me confused (contexts.clear could actually override the model field if providingContext were null).

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+
+	@Override
+	protected void addProvidingAssemblyContext_AssemblyConnectorPropertyDescriptor(Object object) {
+		super.addProvidingAssemblyContext_AssemblyConnectorPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(AssemblyConnector.class, AssemblyContext.class) {
+			@Override
+            protected Collection<?> getValueChoiceTyped(AssemblyConnector object,
+                    List<AssemblyContext> typedList) {
+				// Für Filterschritte private Methoden bauen die predicates zurückggeben, jenachdem welcher Filter bruacht.
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:116]: TODO comments should start with /TODO as they are separately highlighted by most IDEs. Remove if you don't feel they are necessary anymore.

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 7]
```diff
@@ -0,0 +1,341 @@
...
+            				.getEncapsulatedComponent__AssemblyContext()
+            				.getRequiredRoles_InterfaceRequiringEntity()
+            				.stream()
+            				.filter(OperationRequiredRole.class::isInstance)
+                			.map(OperationRequiredRole.class::cast).collect(Collectors.toList());
+            		requiringContextInterfaces = requiringContextRoles.stream()
+            				.map(OperationRequiredRole.class::cast)
+            				.map(OperationRequiredRole::getRequiredInterface__OperationRequiredRole)
+            				.collect(Collectors.toList());
+            		//Schmeiße alle Contexte raus, die keinen Match mit der requiringContextInterfaces Liste haben:
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:134]: Please translate your comments into english.

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 8]
```diff
@@ -0,0 +1,341 @@
...
+					operationRequiredRoles = operationRequiredRoles.stream()
+							.filter(orr -> orr.getRequiredInterface__OperationRequiredRole().isAssignableFrom(providingInterface))
+							.collect(Collectors.toList());
+				}
+            	return operationRequiredRoles;
+            }
+		});
+	}
+
+	private List<AssemblyContext> removeContext(Collection<AssemblyContext> contexts, AssemblyConnector connector, RequiringProviding flag) {
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:290]: This method name is rather confusing. You do not remove the context from the underlying list, but return a copy without it. I would rather go for excludeContext or getCopyWithoutContext

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 9]
```diff
@@ -0,0 +1,341 @@
...
+		AssemblyContext myContext;
+		if (flag == RequiringProviding.REQUIRING) {
+			myContext = connector.getRequiringAssemblyContext_AssemblyConnector();
+		}	else {
+			myContext = connector.getProvidingAssemblyContext_AssemblyConnector();
+		}
+		return contexts.stream().filter(c -> c != myContext).collect(Collectors.toList());
+	}
+	
+	private Predicate<AssemblyContext> contextProvidingRolesAssignableFromRole(AssemblyConnector connector, RequiringProviding flag) {
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:300]: In general, although these are private methods, could you provide some short Java-Doc commentary stating what kind of filter is created by the method?

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java [hunk 10]
```diff
@@ -0,0 +1,341 @@
...
+			Predicate<OperationProvidedRole> predicate = role -> role
+    				.getProvidedInterface__OperationProvidedRole()
+    				.isAssignableFrom(requiredRole.getRequiredInterface__OperationRequiredRole());
+			return context -> context.getEncapsulatedComponent__AssemblyContext()
+					.getProvidedRoles_InterfaceProvidingEntity().stream().filter(OperationProvidedRole.class::isInstance)
+					.map(OperationProvidedRole.class::cast).anyMatch(predicate);
+		}
+	}
+
+	private Predicate<AssemblyContext> contextRequiringRolesAssignableFromRole(AssemblyConnector connector, RequiringProviding flag) {
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/core/composition/provider/AssemblyConnectorItemProvider.java:320]: same here.

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/usagemodel/provider/StartItemProvider.java [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+		super.addPredecessorPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(Start.class, AbstractUserAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(Start object,
+                    List<AbstractUserAction> typedList) {
+            	return new ArrayList<AbstractUserAction>();
+            }
+		});
+	}
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/usagemodel/provider/StartItemProvider.java:32]: same here

File: bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/usagemodel/provider/StopItemProvider.java [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+		super.addSuccessorPropertyDescriptor(object);
+		var decorator = ItemPropertyDescriptorUtils.decorateLastDescriptor(this.itemPropertyDescriptors);
+		decorator.setValueChoiceCalculator(new ValueChoiceCalculatorBase<>(Stop.class, AbstractUserAction.class) {
+            @Override
+            protected Collection<?> getValueChoiceTyped(Stop object,
+                    List<AbstractUserAction> typedList) {
+            	return new ArrayList<AbstractUserAction>();
+            }
+		});
+	}
```
Comment (sdkrach) [inline bundles/org.palladiosimulator.pcm.edit/src/org/palladiosimulator/pcm/usagemodel/provider/StopItemProvider.java:31]: same here

File: tests/org.palladiosimulator.pcm.edit.tests/src/org/palladiosimulator/pcm/edit/provider/TestItemProviderUtilities.java [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+
+	protected static System loadSystem() {
+		
+		Resource.Factory.Registry reg = Resource.Factory.Registry.INSTANCE;
+        Map<String, Object> m = reg.getExtensionToFactoryMap();
+        m.put("system", new XMIResourceFactoryImpl());
+        
+        ResourceSet resSet = new ResourceSetImpl();
+        Resource resource = resSet.getResource(URI
+                .createURI("testmodels/MediaStore-InstantDownloadCache.system"), true);
```
Comment (sdkrach) [inline tests/org.palladiosimulator.pcm.edit.tests/src/org/palladiosimulator/pcm/edit/provider/TestItemProviderUtilities.java:42]: You should put these hard-coded strings into Constants at the top of the class. Or create a constructor which allows you to reuse these utilities by providing a foldername, you can instantiate it to an attribute in your test, giving the concrete folder name)

**Review by sdkrach — CHANGES_REQUESTED** [review]
In principle the changes look good to me. Some minor remarks you find inline.

Furthermore: Please check your source code style settings. There is a mixture of space and tab indentation. Please check the indentation as well. Please use the coding conventions from our github repository (https://github.com/PalladioSimulator/Palladio-Build-CodingConventions)

**Review by rapp-fzi — APPROVED** [review]
temporal approve to trigger build pipeline

**Review by Nicolas-Boltz — DISMISSED** [review]
All comments fixed

**Comment by rapp-fzi** [comment]
Closed this  PR due to migration to new build environment. PR will be re-submitted

**Pull request closed without merging**